torch/csrc/jit/codegen/cuda/lower_index.cpp

```diff
@@ -172,10 +173,28 @@ Val* getGridCommWorkBufferSize(
     }
     buffer_size = SimplifyingIrBuilder::mulExpr(buffer_size, pt_dim);
   }
+
+  // All iteration domains require a separate entry in the buffer for re-entrant
+  // grid reductions.
+  for (auto fl : for_loops) {
+    if (fl->isTrivial()) {
+      continue;
+    }
+    if (fl->iter_domain()->isThread()) {
+      // already accounted for.
+      continue;
+    }
+  
+    buffer_size =
+        SimplifyingIrBuilder::mulExpr(buffer_size, fl->iter_domain()->extent());
+  }
+
   return buffer_size;
 }
 
-Val* getGridSyncBufferSize(const TensorDomain* td) {
+Val* getGridSyncBufferSize(
```

Comment:
No, we shouldn't need to. Because we wait until all iterations are done to start cleaning any of them up. Maybe that's a reason it's slow, I think we should use multiple sync buffers for each reduction!

Reply:
I don't know if it would make a difference, but seems like it's low risk.